### PR TITLE
Replace existing JSON parser with RapidJSON

### PR DESCRIPTION
**Pull request by rok-cesnovar**
#### Submisison Checklist

- [x] Run tests: `./runCmdStanTests.py src/test`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

This PR replaces the existing parser with RapidJSON for faster parsing. With the test made by @mitzimorris I see a 5-fold speedup in an execution of a simple model made to showcase IO speed. The mean of 10 runs for the existing parser 196ms and 38ms for the new parser. I will post the entire results in a PR comment.

There are the following differences with the existing parser:
- for really small numbers (9.999e-1000000000000) the new parser rounds to 0, thje existing one threw
- for integers that are out of int range the parser with try to parse them as doubles (This one is maybe a bit weird, I have a a one-liner fix ready if anyone feels that this is bad.)
- some error messages change, most are more informative but some are maybe less
- the parser does not allow multiple top-level objects (JSON standard doesnt either), the existing one took the first object and ignored any subsequent ones
- there are minor differences with regards to UTF-16 validation  (see tests for example)

#### Intended Effect:

Faster JSON parsing.

#### How to Verify:

Take the model and data from [this](https://github.com/stan-dev/cmdstan/commit/c904e4ca06a3f16b07502429e35b38a727019398) commit, compile the model an run 

```
for i in {1..10}; do echo $i; time (./bym2_read_data sample num_warmup=1 num_samples=1 data file=bym2_big.data.json >/dev/null); echo ""; done
```
Run the following on this branch and on develop. Dont forget to make clean-all in between.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Rok Češnovar

@mitzimorris  provided the speed test. Many thanks for that!

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
Existing parser timings: [existing.txt](https://github.com/stan-dev/cmdstan/files/3627918/existing.txt)
RapidJSON timings: [rapidjson.txt](https://github.com/stan-dev/cmdstan/files/3627919/rapidjson.txt)

A while back I did a memory consumption test for a 2GB input file:
62003922-b11b4380-b11e-11e9-9daa-e13be6310eaf

I also did a test with [optimize on a bernoulli logit glm](https://jeffpollock9.github.io/maximum-likelihood-estimation-with-tensorflow-probability-and-stan-take-2/) for N= 100000 and P = 100, which made a 74MB JSON and 191MB R dump.

Existing parser: 20s
RapidJSON: 2.5s
Rdump: 13s

This is from my Lenovo T450s on Ubuntu 18.04.

File: src/cmdstan/io/json/rapidjson_parser.hpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+namespace cmdstan {
+    namespace json {
+        enum class ParsingState { Idle, Started, End };
+
+        template<typename Handler>
+        struct RapidJSONHandler {
+            explicit RapidJSONHandler(Handler& h) : h_(h),
+                state_(ParsingState::Idle) {
+            }
+            bool check_start() {
```
Comment (rok-cesnovar) [inline src/cmdstan/io/json/rapidjson_parser.hpp]: This check is here because Stan only allow object or arrays on the top level.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 2]
```diff
@@ -215,7 +215,7 @@ TEST(ioJson,jsonParserO4) {
              << 4.01001 
              << "E:arr" 
              << "E:obj" << "E:text";
-  test_parser("{ \"foo\" : 1, \n \"baz\" : [ 2, 3, 4.01001 ] };",
+  test_parser("{ \"foo\" : 1, \n \"baz\" : [ 2, 3, 4.01001 ] }",
```
Comment (rok-cesnovar) [inline src/test/interface/io/json/json_parser_test.cpp]: Nothing is allowed after the top level comment, so I had to delete ";" fromt the test.

**Comment by seantalts** [comment]
Those are some really awesome results, thank you! Excited for this to be in and watch those "100x faster than Stan" benchmark numbers from competing packages drop :P

File: src/test/interface/io/json/json_parser_test.cpp [hunk 3]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
-/*
-TEST(ioJson,jsonParserErr06) {
-  test_exception("[ \"\\uFFFF\" ]",
-                 "unicode escapes not supported\n");
-}
-*/
-
-TEST(ioJson,jsonParserErr06a) {
-  test_exception("[ \"\\uDD1Eabd \" ]",
-                 "illegal unicode values, found low-surrogate, missing high-surrogate\n");
```
Comment (rok-cesnovar) [inline src/test/interface/io/json/json_parser_test.cpp]: This is the test I was telling you about @mitzimorris 
This does not fail with RapidJSON. I can look into this if this is a problem.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 4]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
 TEST(ioJson,jsonParserErr06b) {
   test_exception("[ \"\\uD834abc\" ]",
-                 "illegal unicode values, found high-surrogate, expecting low-surrogate\n");
-}
-
-
-// surrogate pair boundary conditions
-TEST(ioJson,jsonParserErr06c) {
-  test_exception("[ \"\\uDC00\\uDFFFF\" ]",
-              "illegal unicode values, found low-surrogate, missing high-surrogate\n");
```
Comment (rok-cesnovar) [inline src/test/interface/io/json/json_parser_test.cpp]: And the other that doesnt fail.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 5]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
 }
 
 TEST(ioJson,jsonParserErr18) {
   test_exception("[ -1, -2, \"-inf\", ]",
-                 "in array, expecting value\n");
-}
-
-
-TEST(ioJson,jsonParserErr19a) {
-  test_exception("[ 1111111111111111111111111111 ]",
```
Comment (rok-cesnovar) [inline src/test/interface/io/json/json_parser_test.cpp]: This gets parsed as a double and it doesnt throw.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 6]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
-}
-
-TEST(ioJson,jsonParserErr19b) {
-  test_exception("[ -1111111111111111111111111111 ]",
-                 "number exceeds integer range\n");
-}
-
-
-TEST(ioJson,jsonParserErr19c) {
-  test_exception("[ 9.9999999e-1000000000000 ]",
```
Comment (rok-cesnovar) [inline src/test/interface/io/json/json_parser_test.cpp]: This one gets rounded to 0.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 2 above)
Comment (mitzimorris) [reply src/test/interface/io/json/json_parser_test.cpp]: I was a Java/C++ programmer when I wrote those tests - couldn't imagine not putting a semi-colon at the end of something.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 7]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
   test_exception("[ 99.9",
-                 "unexpected end of stream\n");
+                 "Missing a comma or ']' after an array element.\n");
 }
 
 
 TEST(ioJson,jsonParserErr03) {
   test_exception("[ 000.005 ]",
-                 "zero padded numbers not allowed\n");
+                 "Missing a comma or ']' after an array element.\n");
```
Comment (mitzimorris) [inline src/test/interface/io/json/json_parser_test.cpp]: if the error here is that 0-padded numbers are problematic, this message doesn't make sense.  are 0-padded numbers problematic?  can the parser handle '022'?

File: src/test/interface/io/json/json_parser_test.cpp [hunk 8]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
 
 TEST(ioJson,jsonParserErr02c) {
   test_exception("[ 99.9",
-                 "unexpected end of stream\n");
+                 "Missing a comma or ']' after an array element.\n");
 }
 
 
 TEST(ioJson,jsonParserErr03) {
   test_exception("[ 000.005 ]",
```
Comment (mitzimorris) [inline src/test/interface/io/json/json_parser_test.cpp]: are these messages all determined by rapidJSON's parser?  if so, that's great - some of these messages are better than those from the ad-hoc parser, some of them are wierder.  overall, this is fine.

File: src/test/interface/io/json/json_parser_test.cpp [hunk 9]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
-                 "zero padded numbers not allowed\n");
+                 "Missing a comma or ']' after an array element.\n");
 }
 
 TEST(ioJson,jsonParserErr04) {
   test_exception("[ 1. ]",
-                 "expected digit after decimal\n");
+                 "Miss fraction part in number.\n");
```
Comment (mitzimorris) [inline src/test/interface/io/json/json_parser_test.cpp]: should be "missing" - again, if this under our control?

File: src/test/interface/io/json/json_parser_test.cpp [hunk 10]
```diff
@@ -363,231 +363,187 @@ TEST(ioJson,jsonParserErr01) {
...
-TEST(ioJson,jsonParserErr06) {
-  test_exception("[ \"\\uFFFF\" ]",
-                 "unicode escapes not supported\n");
-}
-*/
-
-TEST(ioJson,jsonParserErr06a) {
-  test_exception("[ \"\\uDD1Eabd \" ]",
-                 "illegal unicode values, found low-surrogate, missing high-surrogate\n");
+                 "Miss exponent in number.\n");
```
Comment (mitzimorris) [inline src/test/interface/io/json/json_parser_test.cpp]: same as above

File: src/test/interface/io/json/json_parser_test.cpp (hunk 3 above)
Comment (mitzimorris) [reply src/test/interface/io/json/json_parser_test.cpp]: I'm OK with the difference.  let's see if this problem occurs in the wild before spending energy on it.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 4 above)
Comment (mitzimorris) [reply src/test/interface/io/json/json_parser_test.cpp]: OK - see above comment.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 6 above)
Comment (mitzimorris) [reply src/test/interface/io/json/json_parser_test.cpp]: OK

File: src/test/interface/io/json/json_parser_test.cpp (hunk 5 above)
Comment (mitzimorris) [reply src/test/interface/io/json/json_parser_test.cpp]: is that number within double range?  can you craft a number that's bigger than max double value?

File: src/test/interface/io/json/json_parser_test.cpp (hunk 7 above)
Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: Will get to the bottom of this, this one is weird.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 8 above)
Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: They are determined by them but they are easy to change if we want to.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 9 above)
Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: Yes, its easy to change, will do that.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 7 above)
Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: The reason for such a weird message is that the parser expects a comma or ] after the zero ("." is the third option if its a decimal number). 0-padded numbers are problematic and are not allowed per JSON standard. Some parsers do allow it. This one does not. One option of solving this would be to expand the error message to something like 
`"Missing a comma or ']' after an array element or found a zero padded number.\n"`

The other would be to change the guts of the parser.

Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: I added the above idea for now, we can revert it if its not fine.

File: src/test/interface/io/json/json_parser_test.cpp (hunk 5 above)
Comment (rok-cesnovar) [reply src/test/interface/io/json/json_parser_test.cpp]: Yes, 310 ones do it. That throws "Number too big to be stored in double.\n"

**Comment by rok-cesnovar** [comment]
I fixed everything according to the review comments. The error messages are defined in https://github.com/stan-dev/cmdstan/pull/732/files#diff-f7777e1a53c5e51527ca234b226cb7af
and can easily be changed. The changes to this file are the only changes I did to the parser.

**Comment by mitzimorris** [comment]
beautiful!!
tested for myself:

rapidjson:
```
real	0m0.025s
user	0m0.019s
sys	0m0.004s
```

CmdStan 2.20.0:
```
real	0m0.139s
user	0m0.132s
sys	0m0.005s
```

many thanks!!!
